### DISCUSSION TITLE
Pre-hashing Signatures (and Javascript/Python)

### DISCUSSION THREAD

**Discussion by bkgoodman** [jedisct1/libsodium#1143] [Q&A]
The docs say:

> Note: Ed25519ph(m) is intentionally not equivalent to Ed25519(SHA512(m)).
> If, for some reason, you need to prehash the message yourself, use the multi-part crypto_generichash_*() APIs and sign the 512 bit output.

1. Is there any example/sample code on how to do this?
2. _Which hash_ am I supposed to use (to generate signatures compatible with all the other standard APIs)?
3. I _really_ want to do this (sign and verify by hash) in both Python and Javascript - but it doesn't look like the support/bindings are not available to do this in either. Am I wrong on this, or does anyone know of a solution?

**Comment by jedisct1** [comment]
```js
const _sodium = require('libsodium-wrappers');

(async () => {
    await _sodium.ready;
    const sodium = _sodium;

    const kp = sodium.crypto_sign_keypair();

    let st = sodium.crypto_generichash_init(null, 64);
    sodium.crypto_generichash_update(st, sodium.from_string("Hello "));
    sodium.crypto_generichash_update(st, sodium.from_string("world!"));
    let h = sodium.crypto_generichash_final(st, 64);

    const sig = sodium.crypto_sign_detached(h, kp.privateKey);
    console.log(sodium.to_hex(sig));
})();
```

**Comment by jedisct1** [comment]
I've never used libsodium with Python, but it should be similar.
